mava/systems/ppo/sebulba/ff_ippo.py

```diff
@@ -11,3 +11,811 @@
+    for _update in range(config.system.num_updates):
+        inference_time: float = 0
+        storage_time: float = 0
+        env_send_time: float = 0
+
+        # Get the latest parameters from the learner
+        params_queue_get_time_start = time.time()
+        params = params_queue.get()
+        params_queue_get_time.append(time.time() - params_queue_get_time_start)
+
+        # Rollout
+        rollout_time_start = time.time()
+        storage: List = []
+
+        # Loop over the rollout length
+        for _ in range(0, config.system.rollout_length):
+            # Cached for transition
+            cached_next_obs = move_to_device(
+                jnp.stack(next_obs, axis=1)
+            )  # (num_envs, num_agents, ...)
+            cached_next_dones = move_to_device(next_dones)  # (num_envs, num_agents)
+            cashed_action_mask = move_to_device(
+                np.stack(info["actions_mask"])
+            )  # (num_envs, num_agents, num_actions)
+
+            full_observation = Observation(cached_next_obs, cashed_action_mask)
+            # Get action and value
+            inference_time_start = time.time()
+            (
+                action,
+                log_prob,
+                value,
+                key,
+            ) = get_action_and_value(params, full_observation, key)
+
+            # Step the environment
+            inference_time += time.time() - inference_time_start
+            env_send_time_start = time.time()
+            cpu_action = jax.device_get(action)
+            next_obs, next_reward, terminated, truncated, info = env.step(
+                cpu_action.swapaxes(0, 1)
+            )  # (num_env, num_agents) --> (num_agents, num_env)
+            env_send_time += time.time() - env_send_time_start
+
+            # Prepare the data
+            storage_time_start = time.time()
+            next_dones = np.logical_or(terminated, truncated)
+            metrics = jax.tree_util.tree_map(
+                lambda *x: jnp.asarray(x), *info["metrics"]
+            )  # Stack the metrics
+
+            # Append data to storage
+            storage.append(
+                PPOTransition(
+                    done=cached_next_dones,
+                    action=action,
+                    value=value,
+                    reward=next_reward,
+                    log_prob=log_prob,
+                    obs=full_observation,
+                    info=metrics,
+                )
+            )
+            storage_time += time.time() - storage_time_start
+        rollout_time.append(time.time() - rollout_time_start)
+
+        parse_timer = time.time()
+
+        # Prepare data to share with learner
+        # [PPOTransition() * rollout_len] --> PPOTransition[done=(rollout_len, num_envs, num_agents)
+        # , action=(rollout_len, num_envs, num_agents, num_actions), ...]
+        stacked_storage = jax.tree_util.tree_map(lambda *xs: jnp.stack(xs), *storage)
```

Comment:
double check, but I think this will work
```suggestion
        stacked_storage = jax.tree_util.tree_map(jnp.stack, storage)
```

Reply:
Same `jax.tree_util.tree_map` -> `tree.map`